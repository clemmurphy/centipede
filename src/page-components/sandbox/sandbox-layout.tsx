import styled from 'styled-components'
import { useDrag, useDrop } from 'react-dnd'
import { DraggableItemTypes } from 'constants/constants'
import { useState } from 'react'

const CenterEverythingDiv = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  gap: 18px;
  height: 100vh;
  width: 100vw;
`

export function SandboxLayout() {
  return (
    <CenterEverythingDiv>
      <DraggableComponent />
      <DroppableComponent />
      <DraggableComponent2 />
    </CenterEverythingDiv>
  )
}

function DraggableComponent() {
  const [{ isDragging }, drag] = useDrag(() => ({
    type: DraggableItemTypes.CARD,
    item: { value: 2 },
    collect: (monitor) => ({
      isDragging: !!monitor.isDragging(),
    }),
    options: {
      dropEffect: 'copy',
    },
    end: (item, monitor) => {
      const dropResult = monitor.getDropResult()
    },
  }))

  return (
    <div
      ref={drag}
      style={{
        opacity: isDragging ? 0.5 : 1,
        fontSize: 25,
        fontWeight: 'bold',
        cursor: 'move',
        background: 'lightblue',
        padding: 20,
      }}
    >
      <p>+ 2</p>
    </div>
  )
}

function DraggableComponent2() {
  const [{ isDragging }, drag] = useDrag(() => ({
    type: DraggableItemTypes.CARD,
    item: { value: -1 },
    collect: (monitor) => ({
      isDragging: !!monitor.isDragging(),
    }),
    options: {
      dropEffect: 'copy',
    },
    end: (item, monitor) => {
      const dropResult = monitor.getDropResult()
    },
  }))

  return (
    <div
      ref={drag}
      style={{
        opacity: isDragging ? 0.5 : 1,
        fontSize: 25,
        fontWeight: 'bold',
        cursor: 'move',
        background: 'lightblue',
        padding: 20,
      }}
    >
      <p>- 1</p>
    </div>
  )
}

let dropCount = 0

function DroppableComponent() {
  const [{ isOver }, drop] = useDrop(() => ({
    accept: DraggableItemTypes.CARD,
    drop: (props: any, monitor: any) => {
      const item = monitor.getItem()
      dropCount = dropCount + item.value
    },
    collect: (monitor) => ({
      isOver: !!monitor.isOver(),
      canDrop: !!monitor.canDrop(),
    }),
  }))

  const color = isOver ? 'lightgreen' : 'lightpink'

  return (
    <div
      style={{
        height: 200,
        width: 200,
        backgroundColor: color,
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        fontSize: 32,
      }}
      ref={drop}
    >
      <p>{dropCount}</p>
    </div>
  )
}
