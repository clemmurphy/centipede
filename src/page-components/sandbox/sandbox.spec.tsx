import { render } from '@testing-library/react'
import { SandboxLayout } from './sandbox-layout'

describe('Sandbox page', () => {
  it('should render', () => {
    const { baseElement } = render(<SandboxLayout />)

    expect(baseElement).toMatchSnapshot()
  })
})
