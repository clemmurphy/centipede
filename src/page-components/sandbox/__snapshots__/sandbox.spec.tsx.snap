// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sandbox page should render 1`] = `
<body>
  <div>
    <div
      class="sandbox-layout__CenterEverythingDiv-sc-1j7nf6l-0 erNWbK"
    >
      <div>
        <h1>
          Sandbox
        </h1>
        <p>
          This is a sandbox page. You can use it to experiment with new components and features.
        </p>
      </div>
    </div>
  </div>
</body>
`;
